src/main/java/com/hotels/service/tracing/zipkintohaystack/ingresses/kafka/KafkaConsumerLoop.java

```diff
@@ -0,0 +1,130 @@
+        this.fork = fork;
+        this.config = config;
+        this.validator = validator;
+        this.decoder = decoder;
+        this.spansCounter = spansCounter;
+    }
+
+    public void initialize() {
+        this.sourceTopics = config.getSourceTopics();
+        this.pollDurationMs = config.getPollDurationMs();
+
+        String kafkaBrokers = config.getBootstrapServers();
+        int autoCommitIntervalMs = config.getAutoCommitIntervalMs();
+        boolean enableAutoCommit = config.isEnableAutoCommit();
+        int sessionTimeoutMs = config.getSessionTimeoutMs();
+        String autoOffsetReset = config.getAutoOffsetReset();
+
+        this.kafkaConsumer = kafkaConsumer(kafkaBrokers, autoCommitIntervalMs, enableAutoCommit, sessionTimeoutMs, autoOffsetReset);
+    }
+
+    private KafkaConsumer<String, byte[]> kafkaConsumer(String kafkaBrokers, int autoCommitIntervalMs, boolean enableAutoCommit, int sessionTimeoutMs, String autoOffsetReset) {
+        return new KafkaConsumer<>(
+                Map.of(
+                        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBrokers,
+                        ConsumerConfig.GROUP_ID_CONFIG, "pitchfork",
+                        ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, autoCommitIntervalMs,
+                        ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, enableAutoCommit,
+                        ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, sessionTimeoutMs,
+                        ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, autoOffsetReset
+                ),
+                new StringDeserializer(),
+                new ByteArrayDeserializer()
+        );
+    }
+
+    @Override
+    public void run() {
+        try {
+            kafkaConsumer.subscribe(sourceTopics);
+
+            while (true) {
+                consumerLock.readLock().lock();
+                var records = kafkaConsumer.poll(ofMillis(pollDurationMs));
+
+                if (!records.isEmpty()) {
+                    StreamSupport.stream(records.spliterator(), false)
+                            .flatMap((Function<ConsumerRecord<String, byte[]>, Stream<Span>>) record -> decoder.decodeList(record.value())
+                                    .stream())
+                            .filter(validator::isSpanValid)
+                            .peek(span -> spansCounter.increment())
+                            .forEach(span -> fork.processSpan(span)
+                                    .doOnError(throwable -> logger.warn("operation=fetchRecordsFromKafka", throwable))
+                                    .onErrorResume(e -> Mono.empty())
+                                    .blockLast());
+                    // TODO: consider replacing with a reactor KafkaReceiver
+                }
+                consumerLock.readLock().unlock();
+            }
+        } catch (WakeupException exception) {
+            // ignore for shutdown
+        } finally {
+            consumerLock.writeLock().lock();
```

Comment:
I think the locks here are unnecessary. The consumer is not thread safe and I was having issues when sharing it across multiple threads, and that's when these were added.
I since have one consumer per thread but forgot to remove these. I need to do some more tests and read the docs again.

Reply:
Agree, no need to lock here.